### PR TITLE
[DROOLS-7054] Create stub definition of "EfestoContext"

### PR DESCRIPTION
**Pull request by gitgabrio**
@mariofusco @evacchi @danielezonca @tkobayas @tarilabs @pibizza 

Definition of a stub efesto "Context" and "Listener"

This PR
1) define a "stub" EfestoContext that extends  `org.kie.api.runtime.Context`
2) define `EfestoListener` marker interface, to be propagated inside the Context and available between all plugins/implementatios
3) fix artifact naming (every efesto artifact name should start with "efesto-" for consistency)

**JIRA**:
[link](https://issues.redhat.com/browse/DROOLS-7054)

Needed by

https://github.com/kiegroup/drools/pull/4503

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] mandrel</b>
</details>

### CODE REVIEW

**Comment by kie-ci3** [comment]
**(tests) - kogito-runtimes job** `#213` was: **UNSTABLE**
Possible explanation: This should be test failures

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-runtimes/213/display/redirect

**Test results:**
- PASSED: 2812
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-runtimes/213/testReport/org.kie.kogito.codegen.rules/BigRuleSetCodegenTest/test/">org.kie.kogito.codegen.rules.BigRuleSetCodegenTest.test</a></summary>
org.kie.kogito.codegen.rules.RuleCodegenError: <br/>Errors were generated during the code-generation process:<br/>java.lang.IllegalStateException: No data of this type found. Use containsData to check for this first.<br/><br/>	at org.kie.kogito.codegen.rules.DroolsModelBuilder.build(DroolsModelBuilder.java:79)<br/>	at org.kie.kogito.codegen.rules.RuleCodegen.internalGenerate(RuleCodegen.java:110)<br/>	at org.kie.kogito.codegen.core.AbstractGenerator.generate(AbstractGenerator.java:69)<br/>	at org.kie.kogito.codegen.rules.BigRuleSetCodegenTest.test(BigRuleSetCodegenTest.java:48)<br/>	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)<br/>	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)<br/>	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)<br/>	at java.base/java.lang.reflect.Method.invoke(Method.java:566)<br/>	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)<br/>	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)<br/>	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)<br/>	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)<br/>	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)<br/>	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)<br/>	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)<br/>	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)<br/>	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)<br/>	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)<br/>	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)<br/>	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)<br/>	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)<br/>	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)<br/>	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)<br/>	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)<br/>	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)<br/>	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)<br/>	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)<br/>	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)<br/>	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)<br/>	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)<br/>	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)<br/>	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)<br/>	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)<br/>	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)<br/>	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)<br/>	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)<br/>	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)<br/>	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)<br/>	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)<br/>	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)<br/>	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)<br/>	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)<br/>	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)<br/>	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)<br/>	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)<br/>	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)<br/>	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)<br/>	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)<br/>	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)<br/>	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)<br/>	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)<br/>Caused by: java.lang.IllegalStateException: java.lang.IllegalStateException: No data of this type found. Use containsData to check for this first.<br/>	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)<br/>	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)<br/>	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)<br/>	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)<br/>	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)<br/>	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)<br/>	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)<br/>	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919)<br/>	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)<br/>	at java.base/java.util.stream.IntPipeline.reduce(IntPipeline.java:496)<br/>	at java.base/java.util.stream.IntPipeline.max(IntPipeline.java:459)<br/>	at org.drools.model.codegen.execmodel.PackageModel.generateRulesInUnit(PackageModel.java:708)<br/>	at org.drools.model.codegen.execmodel.PackageModel.getRulesSource(PackageModel.java:612)<br/>	at org.drools.model.codegen.execmodel.PackageModelWriter.<init>(PackageModelWriter.java:38)<br/>	at org.drools.model.codegen.project.KogitoPackageSources.dumpSources(KogitoPackageSources.java:68)<br/>	at org.drools.model.codegen.execmodel.processors.SourceCodeGenerationPhase.process(SourceCodeGenerationPhase.java:51)<br/>	at org.drools.compiler.builder.impl.processors.IteratingPhase.process(IteratingPhase.java:54)<br/>	at org.drools.model.codegen.execmodel.processors.ModelMainCompilationPhase.process(ModelMainCompilationPhase.java:103)<br/>	at org.drools.model.codegen.execmodel.ModelBuilderImpl.doSecondBuildStep(ModelBuilderImpl.java:112)<br/>	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:120)<br/>	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:104)<br/>	at org.kie.kogito.codegen.rules.DroolsModelBuilder.build(DroolsModelBuilder.java:73)<br/>	... 73 more<br/>Caused by: java.lang.IllegalStateException: No data of this type found. Use containsData to check for this first.<br/>	at com.github.javaparser.ast.Node.getData(Node.java:505)<br/>	at com.github.javaparser.ast.CompilationUnit.getPrinter(CompilationUnit.java:154)<br/>	at com.github.javaparser.ast.Node.lambda$getPrinter$1(Node.java:199)<br/>	at java.base/java.util.Optional.map(Optional.java:265)<br/>	at com.github.javaparser.ast.Node.getPrinter(Node.java:199)<br/>	at com.github.javaparser.ast.Node.toString(Node.java:322)<br/>	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)<br/>	at java.base/java.util.TreeMap$ValueSpliterator.forEachRemaining(TreeMap.java:2890)<br/>	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)<br/>	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)<br/>	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952)<br/>	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926)<br/>	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)<br/>	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)<br/>	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)<br/>	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)<br/>	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)<br/>	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)<br/>	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)<br/>
</details>

**Comment by gitgabrio** [comment]
Kogito Apps failure unrelated
kogito-runtimes failure unrelated
SonarCloud failure not relevant

File: efesto/efesto-core/efesto-runtime-manager/efesto-runtime-manager-api/src/main/java/org/kie/efesto/runtimemanager/api/utils/SPIUtils.java [hunk 1]
```diff
@@ -41,17 +43,18 @@ private SPIUtils() {
...
         logger.debug("getKieRuntimeService {} {}", input, refresh);
-        return findAtMostOne(getKieRuntimeServices(refresh), service -> service.canManageInput(input, memoryCompilerClassLoader),
-                (s1, s2) -> new KieRuntimeServiceException("Found more than one compiler services: " + s1 + " and " + s2));
+        return findAtMostOne(getKieRuntimeServices(refresh), service -> service.canManageInput(input,
+                                                                                               memoryCompilerClassLoader),
+                             (s1, s2) -> new KieRuntimeServiceException("Found more than one compiler services: " + s1 + " and " + s2));
     }
 
-    public static List<KieRuntimeService> getKieRuntimeServices(boolean refresh) {
+    public static <S, U, T extends EfestoInput<S>, E extends EfestoOutput<U>> List<KieRuntimeService<S, U, T, E>> getKieRuntimeServices(boolean refresh) {
```
Comment (evacchi) [inline efesto/efesto-core/efesto-runtime-manager/efesto-runtime-manager-api/src/main/java/org/kie/efesto/runtimemanager/api/utils/SPIUtils.java]: this many type parameters may become unwieldy as we carry forward. Let's keep an eye on those.

**Review by evacchi — APPROVED** [review]
I left a minor comment on the use of type parameters that we do not have to address right now, just a consideration for the future. LGTM

File: efesto/efesto-core/efesto-runtime-manager/efesto-runtime-manager-api/src/main/java/org/kie/efesto/runtimemanager/api/utils/SPIUtils.java (hunk 1 above)
Comment (gitgabrio) [reply efesto/efesto-core/efesto-runtime-manager/efesto-runtime-manager-api/src/main/java/org/kie/efesto/runtimemanager/api/utils/SPIUtils.java]: :+1: 
I add all of them to remove warnings from Intellij all over the code - even downward. I'm not really opinionated about that.
We can decide if
1) is actualy needed or just an optional
2) if it is needed, we can decide to create a sort of "Container" type that hides all of them (just an idea)
3) it is (somehow) related to a modification that @mariofusco would like to do, but it has to be confirmed with pmml to see if it break the design or not

Comment (mariofusco) [reply efesto/efesto-core/efesto-runtime-manager/efesto-runtime-manager-api/src/main/java/org/kie/efesto/runtimemanager/api/utils/SPIUtils.java]: @gitgabrio @evacchi In general I believe that there is an abuse of generics. In some places they don't add much value (an extensive use of the `<?>` wildcard always rings a bell for me) and they only become a further burden to make the compiler and the IDE happy. Of course this is not very related with this pull request and I don't consider a blocker for it. Anyway I agree that it is something that we need to review as soon as possible.

Comment (gitgabrio) [reply efesto/efesto-core/efesto-runtime-manager/efesto-runtime-manager-api/src/main/java/org/kie/efesto/runtimemanager/api/utils/SPIUtils.java]: > ... they only become a further burden to make the compiler and the IDE happy


Perfectly agree (exactly the reason I add them) :smile:   No problem for me to remove them all even in this PR

**Comment by gitgabrio** [comment]
@mariofusco @evacchi 
Since I had the faint impression you did not like that generics very much, I removed them

**Comment by gitgabrio** [comment]
@danielezonca @mariofusco Hi! Could you please merge this, squashing commits? 
We discussed the "Context/Listener" adding; for the moment being are just "marker interface" needed inside pmml that - in turns - is the litmus test of all the refactoring.
Thanks!

**Review by mariofusco — APPROVED** [review]

**Pull request merged**